Source/ACE/Network/NetworkSession.cs

```diff
@@ -12,110 +13,124 @@ namespace ACE.Network
 {
     public class NetworkSession
     {
+        private readonly Object lockObject = new Object();
+
         private const int minimumTimeBetweenBundles = 5; // 5ms
         private const int timeBetweenTimeSync = 20000; // 20s
         private const int timeBetweenAck = 2000; // 2s
 
-        private Session session;
-        private NetworkBundle currentBundle;
-        private DateTime nextResync;
-        private DateTime nextAck;
-        private DateTime nextSend;
+        private readonly Session session;
+
+        private NetworkBundle currentBundle = new NetworkBundle();
+
+        private DateTime nextResync = DateTime.UtcNow;
+        private DateTime nextAck = DateTime.UtcNow;
+        private DateTime nextSend = DateTime.UtcNow;
         private bool sendAck = false;
         private bool sendResync = false;
         private uint lastReceivedSequence = 0;
-        private ConcurrentDictionary<uint /*seq*/, ServerPacket> CachedPackets { get; } = new ConcurrentDictionary<uint, ServerPacket>();
-        private ConcurrentQueue<ServerPacket> PacketQueue { get; } = new ConcurrentQueue<ServerPacket>();
 
-        public SessionConnectionData ConnectionData { get; private set; }
+        private readonly Dictionary<uint /*seq*/, ServerPacket> cachedPackets = new Dictionary<uint /*seq*/, ServerPacket>();
+
+        private readonly Queue<ServerPacket> packetQueue = new Queue<ServerPacket>();
+
+        public readonly SessionConnectionData ConnectionData = new SessionConnectionData();
 
         public NetworkSession(Session session)
         {
             this.session = session;
-            ConnectionData  = new SessionConnectionData();
-            currentBundle = new NetworkBundle();
-            nextSend = DateTime.UtcNow;
-            nextResync = DateTime.UtcNow;
-            nextAck = DateTime.UtcNow;
         }
 
         public void EnqueueSend(params GameMessage[] messages)
         {
-            currentBundle.EncryptedChecksum = true;
-            foreach (var message in messages)
-                currentBundle.Enqueue(message);
+            lock (lockObject)
+            {
+                currentBundle.EncryptedChecksum = true;
+
+                foreach (var message in messages)
+                    currentBundle.Enqueue(message);
+            }
         }
 
         public void EnqueueSend(params ServerPacket[] packets)
         {
-            foreach (var packet in packets)
-                PacketQueue.Enqueue(packet);
+            lock (lockObject)

```

Comment:
this needs to be a different/new lock/mutex - "packetQueueLock" or "packetQueueMutex" would both be good.